rust/lance-core/src/utils/mask.rs

```diff
@@ -264,37 +268,335 @@ impl std::ops::BitOr for RowIdMask {
+    /// for each entry:
+    ///   * u32: fragment_id
+    ///   * u32: bitmap size
+    ///   * [u8]: bitmap
+    /// If bitmap size is zero then the entire fragment is selected.
+    pub fn serialize_into<W: Write>(&self, mut writer: W) -> Result<()> {
+        writer.write_u32::<byteorder::LittleEndian>(self.inner.len() as u32)?;
+        for (fragment, set) in &self.inner {
+            writer.write_u32::<byteorder::LittleEndian>(*fragment)?;
+            if let Some(set) = set {
+                writer.write_u32::<byteorder::LittleEndian>(set.serialized_size() as u32)?;
+                set.serialize_into(&mut writer)?;
+            } else {
+                writer.write_u32::<byteorder::LittleEndian>(0)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Deserialize the set from the given buffer
+    pub fn deserialize_from<R: Read>(mut reader: R) -> Result<Self> {
+        let num_entries = reader.read_u32::<byteorder::LittleEndian>()?;
+        let mut inner = BTreeMap::new();
+        for _ in 0..num_entries {
+            let fragment = reader.read_u32::<byteorder::LittleEndian>()?;
+            let bitmap_size = reader.read_u32::<byteorder::LittleEndian>()?;
+            if bitmap_size == 0 {
+                inner.insert(fragment, None);
+            } else {
+                let mut buffer = vec![0; bitmap_size as usize];
+                reader.read_exact(&mut buffer)?;
+                let set = RoaringBitmap::deserialize_from(&buffer[..])?;
+                inner.insert(fragment, Some(set));
+            }
+        }
+        Ok(Self { inner })
+    }
+}
+
+impl std::ops::BitOr<Self> for RowIdTreeMap {
+    type Output = Self;
+
+    fn bitor(mut self, rhs: Self) -> Self::Output {
+        for (fragment, rhs_set) in &rhs.inner {
+            match self.inner.get_mut(fragment) {
+                None => {
+                    self.inner.insert(*fragment, rhs_set.clone());
+                }
+                Some(None) => {
+                    // If the fragment is already selected then there is nothing to do
+                }
+                Some(Some(lhs_set)) => {
+                    if let Some(rhs_set) = rhs_set {
+                        *lhs_set |= rhs_set;
+                    }
+                }
+            }
+        }
+        self
+    }
+}
+
+impl std::ops::BitAnd<Self> for RowIdTreeMap {
+    type Output = Self;
+
+    fn bitand(mut self, rhs: Self) -> Self::Output {
+        // Remove fragment that aren't on the RHS
+        self.inner
+            .retain(|fragment, _| rhs.inner.contains_key(fragment));
+
+        // For fragments that are on the RHS, intersect the bitmaps
+        for (fragment, mut lhs_set) in &mut self.inner {
+            match (&mut lhs_set, rhs.inner.get(fragment)) {
+                (_, None) => {} // Already handled by retain
+                (_, Some(None)) => {
+                    // Everything selected on RHS, so can leave LHS untouched.
+                }
```

Comment:
this behavior is a little weird. do we ever hit this branch?

Reply:
Why is this weird?

Reply:
I would usually assume if we do and, the element is removed

Reply:
`None` means fragment isn't selected, `Some(None)` means all rows in the fragment is selected. I suppose it's maybe confusing to use `Option` here. Maybe I should refactor to instead:

```rust
enum FragmentSelection {
   Full,
   Bitmap(RoaringBitmap)
}
```

Would that be clearer?